### ISSUE TITLE
[RFC] Introducing a new `--log-level` flag

### ISSUE THREAD

**Issue by yihuaf** [containers/youki#2028]
Currently, there are a few ways we can manage the log level in `youki`.

- Through the `YOUKI_LOG_LEVEL` env variable.
I believe this copying what traditionally rust logging library does. Both the `env-logger` we used previously and `tracing` has native interface to parse log level from the env variable.

- Through the `debug` build.
Currently, if `youki` is build with debug build, the default log level is automatically to `debug`

-Through the `--debug` flag.
Currently, both `runc` and us have a `debug` flag where we turn the log level to `debug`.

Up until now, these 3 options have served us well. However, I would like to introducing a new level `trace` that is more verbose than `debug`. For example, currently I am trying to debug `seccomp` and I need to log a lot of information on exactly what is going on with each seccomp rules addes and translated. Docker by default passes in 200+ seccomp rules in the profiles, so the logs can become very spammy, even by `debug` level standards. I don't think it is good devx to turn these spammy logs on by default. So I believe putting these logs behind `trace` level makes sense.

With this new level, the only way to switch the level during runtime is through the `YOUKI_LOG_LEVEL`. Both `--debug` flag and debug build will not help in this case. However, managing env variable is a pain when using `youki` under docker or containerd, because the config files for those only specify the binary path and args, not env. We can certainly write our own wrapper to add the env variable, but that is a pain and the original inspiration for us to add the `--debug` flag.

One potential solution:

We introduce a runtime `--log-level` flag. This should be easier to use compared to all existing options and more flexible. Also, a `log-level` flag is common is all kinds of CLI.

The potential question is what should we do with the existing options. Should we create a hierarchy where we clearly states one option will override the other option. Or we can reject setting `--debug` and `--log-level` together. Or we can remove the `--debug` flag and in favor of the new `--log-level debug`.

@containers/youki-maintainers Any opinions/ideas/comments?

Another potentially viable option is to turn the default level to `trace`, but I am not too happy with this because potentially as we add more `trace` level logging, it will become way too spammy.

**Comment by utam0k** [comment]
If this is to be implemented, the ENV style should be eliminated.
Too many ways of doing things is not a good thing. It is confusing.

**Comment by utam0k** [comment]
I agree with the introduction of log-level. Instead, let's document how to change log-level in Docker/Podman/K8s.

**Comment by YJDoc2** [comment]
Hey, I agree with @utam0k , we should only keep one way to set log level. I think we should eliminate the env var, setting is is not always possible. What I can suggest is that let use introduce the log-level flag as you suggested, and set the default log level for debug builds to trace. We would need to consider how this might affect the external integration tests that we run, because some of them parse logs to do assertions.

**Comment by yihuaf** [comment]
> Also note that neither of these two have a --log-level flag, so for our tests, if we pass it, these might potentially error.

Yes, this is an issue for us since our integration test runs for both `runc` and `youki`. Before, we use environment variable to set the log level to `error` so our test passes. I would propose we set the default log-level to release build to be `error` instead of `warning`. In this way, we match with `runc` to not log `warning` as well.

> One thing that we should consider is that runc and crun both have --debug flag, so if someone is trying youki as drop-in for these, completely removing the debug will break.

I think we can keep the `--debug` for the compatibility reason listed here, but I would make the new `--log-level` override the `--debug` flag. Since `log-level` flag is very explicit, I doubt anyone would do `--log-level error` with `--debug` together and expect the log level to be debug. I think this is better than error out if both flags are set.

On the other hand, I would keep the log level of debug build to `debug`. `trace` can be very spammy and not needed for most of the time.

In short:

- Propose the default log level for release to `error`
- Keep the `--debug` flag and it sets the log level to debug
- Introduce `--log-level` flag to set the log level
-  `--debug` flag will be ignored if `--log-level` flag is also set.
- Keep the default log level for debug build to `debug`
